components/brave_wallet/browser/solana_keyring.cc

```diff
@@ -57,4 +67,100 @@ std::string SolanaKeyring::GetAddressInternal(HDKeyBase* hd_key_base) const {
   return hd_key->GetBase58EncodedPublicKey();
 }
 
+// Create a valid program derived address without searching for a bump seed.
+// https://docs.rs/solana-program/latest/solana_program/pubkey/struct.Pubkey.html#method.create_program_address
+// static
+absl::optional<std::string> SolanaKeyring::CreateProgramDerivedAddress(
+    const std::vector<std::vector<uint8_t>>& seeds,
+    const std::string& program_id) {
+  const std::string pda_marker = "ProgramDerivedAddress";
+
+  std::vector<uint8_t> program_id_bytes;
+  if (!Base58Decode(program_id, &program_id_bytes, kSolanaPubkeySize))
+    return absl::nullopt;
+
+  if (seeds.size() > kMaxSeeds) {
+    return absl::nullopt;
+  }
+
+  for (const auto& seed : seeds) {
+    if (seed.size() > kMaxSeedLen) {
+      return absl::nullopt;
+    }
+  }
+
+  std::vector<uint8_t> buffer;
+  for (const auto& seed : seeds) {
+    buffer.insert(buffer.end(), seed.begin(), seed.end());
+  }
+
+  buffer.insert(buffer.end(), program_id_bytes.begin(), program_id_bytes.end());
+  buffer.insert(buffer.end(), pda_marker.begin(), pda_marker.end());
+
+  auto hash_array = crypto::SHA256Hash(buffer);
+  std::vector<uint8_t> hash_vec(hash_array.begin(), hash_array.end());
+
+  // Invalid because program derived addresses have to be off-curve.
+  if (bytes_are_curve25519_point(
+          rust::Slice<const uint8_t>{hash_vec.data(), hash_vec.size()})) {
+    return absl::nullopt;
+  }
+
+  return Base58Encode(hash_vec);
+}
+
+// Find a valid program derived address and its corresponding bump seed.
+// https://docs.rs/solana-program/latest/solana_program/pubkey/struct.Pubkey.html#method.find_program_address
```

Comment:
I took a look at the bump seed design described in these documents and I'm fairly certain this will be fine. We'll face the same considerations as what is highlighted in the documents in the fact that we'll face the same localized entropy warnings as what they highlighted in their documents, but I'm not seeing any way around this other than to increate the size of the seeds which would likely cause issues between this implementation and the solana_program implementation. They also expect 8 bit seeds.